tools/fft.c

```diff
@@ -0,0 +1,381 @@
+	for (unsigned i = 0; i < out_length; i++) {
+		printf("%"PRIu64",%u,%f\n", output_index, channel->channel,
+				channel->welch_output[i]);
+	}
+	output_index++;
+}
+
+int add_sample_to_channel(struct channel_data *channel, double sample)
+{
+	unsigned full_windows = 0;
+	struct sample_window *window;
+	if (channel->sample_index > 0
+			&& channel->sample_index % channel->new_window_interval == 0) {
+		struct sample_window *w = create_sample_window(channel);
+		if (w == NULL) {
+			return -errno;
+		}
+		if (!pfifo_write(channel->windows, w)) {
+			return -1;
+		}
+		channel->sample_index = 0;
+	}
+
+	for (unsigned i = 0; pfifo_peek_back(channel->windows, i,
+				(void**) &window); i++) {
+		full_windows += add_sample_to_window(window, sample);
+	}
+
+	if (full_windows == channel->welch_window_count) {
+		welch_method(channel);
+		print_welch_output(channel);
+
+		if (!pfifo_read(channel->windows, (void**) &window)) {
+			return -1;
+		}
+		// TODO: centralise calculation of window_value
+		subtract_from_welch_average(channel, channel->window_length, window);
+		destroy_sample_window(window);
+	}
+
+
+	channel->sample_index++;
+	return 0;
+}
+
+int init_channel(struct channel_data *channel,
+		const bl_msg_channel_conf_t *msg, uint32_t window_count)
+{
+	memset(channel, 0, sizeof(*channel));
+	channel->channel = msg->channel;
+
+	/* I have decided to always use 50% overlap
+	 * If the window length in samples is even-numbered, there'd be 1
+	 * partly-full window when we have window_count full windows.
+	 * With odd-numbered window length, there's a second newly-started window
+	 */
+	channel->windows_capacity = window_count + 2;
+	channel->welch_window_count = window_count;
+	channel->windows = pfifo_create(channel->windows_capacity);
+	if (channel->windows == NULL) {
+		return -errno;
+	}
+	return 0;
+}
+
+int init_channel_samples(struct channel_data *channel,
+		uint32_t window_length_samples)
+{
+	struct sample_window *window;
+	channel->window_length = window_length_samples;
+	channel->new_window_interval = window_length_samples / 2;
+	if (!channel->windows) {
+		fprintf(stderr, "Failed to initialise channel. "
+				"Perhaps a channel config message is missing from the input "
+				"stream\n");
+		return -1;
+	}
+	window = create_sample_window(channel);
+	if (window == NULL) {
+		return -errno;
+	}
+	if (!pfifo_write(channel->windows, window)) {
+		return -1;
+	}
+	return 0;
+}
+
+void destroy_channel(struct channel_data *channel)
+{
+	struct sample_window *window;
+	if (channel->welch_output != NULL) {
+		free(channel->welch_output);
+	}
+	// Destroy all existing windows
+	while(pfifo_read(channel->windows, (void**) &window)) {
+		destroy_sample_window(window);
+	}
+	pfifo_destroy(channel->windows);
+}
+
+int read_stream(uint32_t window_length, uint16_t window_count)
+{
+	union bl_msg_data msg; // message for reading into
+	struct channel_data channels[BL_ACQ__SRC_COUNT]; //One per possible channel
```

Comment:
With the source/channel split now on master, you should rebase and:

```suggestion
	struct channel_data channels[BL_ACQ_CHANNEL_COUNT]; //One per possible channel
```
It comes from `../../acq/channel.h`.

I'd initialise it too, since initialised stuff is easier to clean up:

```suggestion
	struct channel_data channels[BL_ACQ_CHANNEL_COUNT] = {0}; //One per possible channel
```